### PR TITLE
feat(prefetch): add sibling prefetch with coalescing batch reader

### PR DESCRIPTION
**Pull request by hooyao**
## Summary

- **SpanSelector**: picks the densest contiguous window of up to `PrefetchMaxFiles` siblings centered on the trigger file, enforcing a fill-ratio threshold to avoid reading large holes between sparse entries
- **Coalescing batch reader**: opens the ZIP archive once, seeks to the span start, and linearly decompresses wanted files while discarding hole bytes — eliminating repeated seeks for sequential access patterns
- **Smart triggering**: prefetch fires only on cold reads (`wasCached == false`) and optionally on directory listings, with a per-directory in-flight guard preventing duplicate concurrent scans
- **Zero redundant I/O**: `ContainsKey` (lock-free) filters already-cached entries before opening the file; if all span entries are warm, the scan is skipped entirely
- **Independent controls**: `PrefetchOnRead` and `PrefetchOnListDirectory` can be toggled separately; `PrefetchEnabled` is the master switch

## New configuration (`Cache` section)

| Key | Default | Description |
|-----|---------|-------------|
| `PrefetchEnabled` | `true` | Master on/off switch |
| `PrefetchOnRead` | `true` | Trigger on cold file reads |
| `PrefetchOnListDirectory` | `true` | Trigger on directory listings |
| `PrefetchFileSizeThresholdMb` | `10` | Max file size eligible for prefetch |
| `PrefetchMaxFiles` | `20` | Max siblings in one span |
| `PrefetchMaxDirectoryFiles` | `300` | Candidate cap before nearest-offset trim |
| `PrefetchFillRatioThreshold` | `0.80` | Min density to accept a span |

## Test plan

- [x] `SpanSelectorTests` — unit tests for window selection, fill-ratio shrinking, edge cases
- [x] `PrefetchIntegrationTests` — end-to-end: read triggers prefetch, sibling served from cache, re-trigger is a no-op
- [x] `FileContentCacheTests` — `ContainsKey` coverage for memory and disk tiers
- [x] All 347 tests passing

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/ZipDrive.Domain.Tests/PrefetchIntegrationTests.cs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    public async Task FindFilesAsync_TriggersPrefetch_SiblingsWarmedBeforeIndividualReads()
+    {
+        // List the directory — this should trigger prefetch fire-and-forget
+        var listing = await _vfs.ListDirectoryAsync("images.zip/img");
+
+        listing.Should().HaveCount(5, "directory has 5 image files");
+
+        // Give the background prefetch time to complete
+        await Task.Delay(500);
+
```
[the remark this answers is not in the document]
Comment (hooyao) [reply tests/ZipDrive.Domain.Tests/PrefetchIntegrationTests.cs:163]: Fixed in c19b42a. Replaced the fixed `Task.Delay(500)` with a bounded polling loop (`200 × 10 ms`) that exits as soon as `_fileCache.EntryCount` reaches the expected threshold, making the tests deterministic without a hard timeout.

File: src/ZipDrive.Application/Services/SpanSelector.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
+using ZipDrive.Domain.Models;
+
+namespace ZipDrive.Application.Services;
+
+/// <summary>
+/// Pure span selection algorithm for sibling prefetch.
+/// Selects a contiguous window of sibling entries around a trigger file,
+/// maximising useful bytes relative to span bytes (fill ratio).
+/// </summary>
+internal static class SpanSelector
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/SpanSelector.cs:10]: Fixed in c19b42a. Added `[assembly: InternalsVisibleTo("ZipDrive.Domain.Tests")]` to `ZipDrive.Application.csproj` so the test assembly can access `internal` types like `SpanSelector` directly.

File: src/ZipDrive.Application/Services/PrefetchPlan.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+using ZipDrive.Domain.Models;
+
+namespace ZipDrive.Application.Services;
+
+/// <summary>
+/// Describes a contiguous span of sibling ZIP entries selected for prefetch
+/// in a single sequential read.
+/// </summary>
+internal sealed class PrefetchPlan
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/PrefetchPlan.cs:9]: Fixed in c19b42a. `PrefetchPlan` is now accessible to the test assembly via the `InternalsVisibleTo` declaration added to `ZipDrive.Application.csproj`.

File: src/ZipDrive.Infrastructure.Caching/FileContentCache.cs [hunk 4]
```diff
@@ -135,6 +135,38 @@ public async Task<int> ReadAsync(
...
+            {
+                Value = capturedStream,
+                SizeBytes = entry.UncompressedSize,
+                OnDisposed = () => ValueTask.CompletedTask
+            });
+
+        // BorrowAsync stores the entry (thundering herd protection: no-op if already cached).
+        // Immediate dispose releases RefCount to 0 — entry stays cached but eviction-eligible.
+        using ICacheHandle<Stream> handle = await cache.BorrowAsync(
+            cacheKey, _defaultTtl, factory, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.Caching/FileContentCache.cs:167]: Fixed in 6a5222a. Added a `factoryExecuted` flag in `WarmAsync`: the factory sets it to `true` when invoked. In a `finally` block, if the flag is still `false` (cache-hit path where the factory was never called), the provided `decompressedStream` is disposed immediately, preventing the leak.

File: src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs [hunk 5]
```diff
@@ -222,6 +273,291 @@ public VfsVolumeInfo GetVolumeInfo()
...
+
+                // Read FileNameLength (offset 26) and ExtraFieldLength (offset 28) — little-endian
+                ushort fileNameLen = BinaryPrimitives.ReadUInt16LittleEndian(headerBuffer.AsSpan(FileNameLengthOffset, 2));
+                ushort extraLen = BinaryPrimitives.ReadUInt16LittleEndian(headerBuffer.AsSpan(28, 2));
+                int variableHeaderLen = fileNameLen + extraLen;
+
+                // Skip filename + extra field
+                if (variableHeaderLen > 0)
+                {
+                    long skipped = await DiscardBytesAsync(zipStream, variableHeaderLen, discardBuffer, ct).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs:455]: Fixed in 6a5222a. Added validation of the local-file-header signature (`0x04034b50`) immediately after reading the 30-byte fixed header. A mismatch logs a warning and breaks out of the sequential scan, preventing silent corruption when reading past the last entry or encountering a corrupted archive.

File: src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs [hunk 6]
```diff
@@ -222,6 +273,291 @@ public VfsVolumeInfo GetVolumeInfo()
...
+                // Read compressed data into a pooled buffer
+                byte[] compressedBuffer = ArrayPool<byte>.Shared.Rent((int)compressedSize);
+                try
+                {
+                    int compRead = await ReadExactAsync(zipStream, compressedBuffer, 0, (int)compressedSize, ct).ConfigureAwait(false);
+                    currentPos += compRead;
+                    bytesRead += compRead;
+
+                    // Decompress into MemoryStream
+                    MemoryStream decompressed = new((int)wantedEntry.UncompressedSize);
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs:484]: Fixed in 6a5222a. Added overflow guards before both `(int)` casts: entries with `compressedSize > int.MaxValue` or `UncompressedSize > int.MaxValue` are skipped (discard compressed bytes + continue to next entry) rather than silently truncating.

File: src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs [hunk 7]
```diff
@@ -170,8 +206,23 @@ public async Task<int> ReadFileAsync(string path, byte[] buffer, long offset, Ca
 
         // Delegate to file content cache — it owns extraction, caching, and byte retrieval
         string cacheKey = $"{archive.VirtualPath}:{internalPath}";
-        return await _fileContentCache.ReadAsync(
+        bool wasCached = _fileContentCache.ContainsKey(cacheKey);
+        if (!wasCached)
+            _logger.LogInformation("Read (miss): {CacheKey} offset={Offset}", cacheKey, offset);
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs:211]: Fixed in 6a5222a. The per-read cache hit/miss log was downgraded from `LogInformation` to `LogDebug`, keeping the hot path silent at the default log level.

File: tests/ZipDrive.Domain.Tests/PrefetchIntegrationTests.cs [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        };
+
+        await _fileCache.WarmAsync(entry, "manual-key", new MemoryStream(expectedData));
+
+        byte[] readBuffer = new byte[expectedData.Length];
+        // Read directly from the cache (not through VFS)
+        // Verify the entry is in cache and RefCount is 0
+        _fileCache.EntryCount.Should().BeGreaterThanOrEqualTo(1);
+        _fileCache.BorrowedEntryCount.Should().Be(0, "handle was disposed immediately after WarmAsync");
+    }
```
[the remark this answers is not in the document]
Comment (hooyao) [reply tests/ZipDrive.Domain.Tests/PrefetchIntegrationTests.cs:222]: Fixed in c19b42a. The test now calls `_fileCache.ReadAsync(...)` after warming and asserts both the returned byte count and the exact byte content, confirming that the warmed stream is the one served on a cache hit.
